### DISCUSSION TITLE
podman chown failed with large UID and GID

### DISCUSSION THREAD

**Discussion by x2byue** [containers/podman#22296] [General]
I am doing an application containerization from RHEL 8.9 to OpenShift. My DEV environment is podman 4.61 on RHEL 8.9. Since the application and its huge data are owned by a user with uid=1293208 gid=1293060 on RHEL, I created the same UID and GID in the dockerfile. When creating a directory /opt/dmap and RUN chown -R uid:gid /opt/dmap, the chown failed with error: chown: changing ownership of '/opt/dmap': Invalid argument. But if using uid=10001, gid=1002, the chown works. Is there anyway to make the chown working for larg uid and gid?

**Comment by eriksjolund** [comment]
Interesting problem. I first tried to write a Containerfile to reproduce the error message that you see:
```
chown: changing ownership of '/opt/dmap': Invalid argument
```

I'm assuming you are running rootless Podman. 

| experiment | result | comment |
| --                 |   --     |      ----     |
|Experiment 1 | failure | reproducing the same error message that you got |
|Experiment 2 | failure | using __--userns-uid-map__ and __--userns-gid-map__  |
|Experiment 3 | success  |        using __--userns-uid-map__,  __--userns-gid-map__, `remap-uids`, `remap-gids` |
### Experiment 1: failure

1. Create the file _Containerfile_ with the contents
   ```
   FROM fedora
   RUN groupadd  -g 1293060 newgroup
   RUN useradd --no-log-init -u 1293208 -g newgroup newuser
   RUN mkdir -p /opt/dmap
   RUN chown 1293208:1293060 /opt/dmap
   ```
   Regarding the option `--no-log-init`, see [troubleshooting.md](https://github.com/containers/podman/blob/main/troubleshooting.md#6-build-hangs-when-the-dockerfile-contains-the-useradd-command)
2. Build the image with the command
   ```
   podman build --no-cache  .
   ```
   The command prints this output
   ```
   STEP 1/5: FROM fedora
   STEP 2/5: RUN groupadd  -g 1293060 newgroup
   --> b0913da2bf0c
   STEP 3/5: RUN useradd --no-log-init -u 1293208 -g newgroup newuser
   useradd warning: newuser's uid 1293208 outside of the UID_MIN 1000 and UID_MAX 60000 range.
   Setting mailbox file permissions: Invalid argument
   --> d60edcbb812b
   STEP 4/5: RUN mkdir -p /opt/dmap
   --> 4a7bb6057501
   STEP 5/5: RUN chown 1293208:1293060 /opt/dmap
   chown: changing ownership of '/opt/dmap': Invalid argument
   Error: building at STEP "RUN chown 1293208:1293060 /opt/dmap": while running runtime: exit status 1
   ```
   __result:__ the command failed with the error message:
   ```
   chown: changing ownership of '/opt/dmap': Invalid argument
   ```
   In other words, this reproduces the error message that you also got.

### Experiment 2: failure (image commit fails)

1. Build the _Containerfile_ from __Experiment 1__ image with the command
   ```
   podman build --no-cache  \
            --userns-gid-map=0:1:50000 \
            --userns-uid-map=0:1:50000 \
            --userns-uid-map=1293208:0:1  \
            --userns-gid-map=1293060:0:1 .
   ```
   The command prints this output
   ```
   STEP 1/5: FROM fedora
   STEP 2/5: RUN groupadd  -g 1293060 newgroup
   --> 7f98376d7e20
   STEP 3/5: RUN useradd --no-log-init -u 1293208 -g newgroup newuser
   useradd warning: newuser's uid 1293208 outside of the UID_MIN 1000 and UID_MAX 60000 range.
   Error: committing container for step {Env:[DISTTAG=f39container FGC=f39 container=oci PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[useradd --no-log-init -u 1293208 -g newgroup newuser] Flags:[] Attrs:map[] Message:RUN useradd --no-log-init -u 1293208 -g newgroup newuser Heredocs:[] Original:RUN useradd --no-log-init -u 1293208 -g newgroup newuser}: copying layers and metadata for container "659fbfbf1cdf1a8413aa2542aaf76b0485b66190564b2112a788ca9767b608f3": writing blob: adding layer with blob "sha256:1c87ed5690929d1368a52565c112b5ca8780820b8bcb7bddc38cfa52c3c1e3e5": processing tar file(potentially insufficient UIDs or GIDs available in user namespace (requested 1293208:1293060 for /home/newuser): Check /etc/subuid and /etc/subgid if configured locally and run "podman system migrate": lchown /home/newuser: invalid argument): exit status 1
   ```
    __result__: The command failed. All the `RUN` steps succeeded but committing the image failed. In other words Experiment 2 came a bit further than Experiment 1.





### Experiment 3: success (configuring `remap-uids` and `remap-gids`)

Try the storage options `remap-uids` and `remap-gids`

1. `sudo useradd test`
2. `sudo machinectl shell --uid test`
3. `mkdir -p ~/.config/containers`
4. Create the file _~/.config/containers/storage.conf_ with the contents
   ```
   [storage]
   driver = "overlay"
   
   [storage.options]
   remap-uids = "0:0:50000:1293208:50001:1"
   remap-gids = "0:0:50000:1293060:50001:1"
   ```
   For more information, see [manpage storage.conf](https://github.com/containers/podman/blob/main/vendor/github.com/containers/storage/storage.conf)
4. `mkdir ~/ctr`
5. `cd ~/ctr`
6. Create the file _~/ctr/Containerfile_  with the same contents as in __Experiment 1__ 
8. Build the container image
   ```
   podman build --no-cache  \
            --userns-gid-map=0:1:50000 \
            --userns-uid-map=0:1:50000 \
            --userns-uid-map=1293208:0:1  \
            --userns-gid-map=1293060:0:1 .
   ```
   The command prints this output
   ```
   STEP 1/5: FROM fedora
   STEP 2/5: RUN groupadd  -g 1293060 newgroup
   --> dd0f719195ec
   STEP 3/5: RUN useradd --no-log-init -u 1293208 -g newgroup newuser
   useradd warning: newuser's uid 1293208 outside of the UID_MIN 1000 and UID_MAX 60000 range.
   --> eb98fd5e2115
   STEP 4/5: RUN mkdir -p /opt/dmap
   --> d88166e9d7ac
   STEP 5/5: RUN chown 1293208:1293060 /opt/dmap
   COMMIT
   --> e031ead4049c
   e031ead4049c155b20f9ed3cdcff51b5cc9d65e393817a0a201e5028919279da
   ```
   __result:__ success

### Discussion

The number 50000 was chosen rather arbitrarily. In __Experiment 3__ I created a new user because I wanted to avoid having to run `podman system reset` or `podman system migrate`. I'm not exactly sure which command is required after modifying `storage.conf`. It's better to experiment in a new user account.

### About the system

Fedora CoreOS 40.20240331.1.0

```
$ podman --version
podman version 5.0.0
```
Native overlayfs is used:
```
$ podman info -f {{.Store.GraphDriverName}}
overlay
$ podman info -f '{{index .Store.GraphStatus "Native Overlay Diff"}}'
true
```

**Comment by x2byue** [comment]
Thanks a lot for your experiment and comments. Here is my case details:
I have a user s-dmadminua in the host: 
uid=1293208(s-dmadminua) gid=1293060(documentum-ua) groups=1293060(documentum-ua),987(libvirt),1000(fuse),90361(unixusers),1242045(unixrsaexception)

I have a user s-dmadminua in container:
uid=50000(s-dmadminua) gid=50000(documentum-ua) groups=50000(documentum-ua),987(libvirt),1000(fuse),90361(unixusers),1242045(unixrsaexception)

Run the container:
podman run -it --rm --name dctmsvr --hostname dctmsvr --memory 16g --log-level debug \
--uidmap=0:50000:10000  \
--gidmap=0:50000:10000  \
--uidmap=1293208:50000:1  \
--gidmap=1293060:50000:1 \
 cs-base-os /bin/bash

Got the error:
DEBU[0000] Failed to create container dctmsvr(dc3d468e80c7887f30c97913d16d8fed1e36d1b46f042a0013d32903c822d922): the specified UID and/or GID mapping 1293208:50000:1 conflicts with other mappings: invalid mappings specified
Error: creating container storage: the specified UID and/or GID mapping 1293208:50000:1 conflicts with other mappings: invalid mappings specified
DEBU[0000] Shutting down engines

How to identify which uid is conflict?

**Reply by eriksjolund** [reply]
The arguments

* `--uidmap=0:50000:10000` 
* `--uidmap=1293208:50000:1`

can't be used at the same time because both specify a mapping for the 50000th subordinate UID of the unprivileged user. 

The mapping is a one-to-one mapping. It's not possible to map a subordinate UID to two different container UIDs.

### update 1

I updated this comment and corrected a mistake I made.

**Comment by x2byue** [comment]
Here is the new podman run commands and error messages:

podman run -it --rm --name dctmsvr --hostname dctmsvr --memory 16g --log-level debug \
--uidmap=0:50000:10000 \
--uidmap=1293208:50000:1 \
cs-base-os /bin/bash

DEBU[0002] Failed to create container dctmsvr(8ae904f9002d2307e821ec86a37cfe33b0fafc1df597ebc9c66daf6695a4ef11): the specified UID and/or GID mapping 1293208:50000:1 conflicts with other mappings: invalid mappings specified
Error: creating container storage: the specified UID and/or GID mapping 1293208:50000:1 conflicts with other mappings: invalid mappings specified
DEBU[0002] Shutting down engines

**Reply by eriksjolund** [reply]
> podman run -it --rm --name dctmsvr --hostname dctmsvr --memory 16g --log-level debug
> --uidmap=0:50000:10000
> --uidmap=1293208:50000:1
> cs-base-os /bin/bash

The arguments

```
 --uidmap=0:50000:10000
 --uidmap=1293208:50000:1
```

both specify how the 50000th subordinate UID is mapped. This does not work because the mapping needs to be a one-to-one mapping. A subordinate UID can either be mapped to one container UID or not at all.
